packages/ui/src/machines/authenticator/signUp.ts

```diff
@@ -261,7 +261,7 @@ export function createSignUpMachine({ services }: SignUpMachineOptions) {
             get(user, 'username') || get(authAttributes, 'username');
           const { password } = authAttributes;
 
-          await Auth.confirmSignUp(username, code);
+          await services.handleConfirmSignUp({ username, code });
```

Comment:
Ahh, I see the mistake I made – custom handlers like `validateSignUp` were being passed as `services` **but aren't actually XState compatible services** with `(context, event)` signatures:
> https://xstate.js.org/docs/guides/communication.html#invoking-promises

Maybe what we should've done is have 1st class properties like https://github.com/NearHuscarl/flutter_login#installation:

```js
<Authenticator
  initialState="signUp"
  handleSignUp={customSignUp}
  // or
  onSignUp={customSignUp}
  validateSignUp={customValidator)
/>
```

The other problem with our XState design is that we have to pass `services` around to create new machines.  Whereas technically these may make more sense on `context` since they're not actual `services`.

Technically to customers `services` is just a wrapper like `components` around common pieces – we just have to recognize internally that some things are actual `services` vs. custom handler functions.

Reply:
Yeah, that could be something we do. But I thought we didn't want to pollute the authenticator with more props?